UnicornRender/video/source/Primitives.cpp

```diff
@@ -61,13 +65,78 @@ void Primitives::Cube(Mesh& mesh)
 void Primitives::Quad(Mesh& mesh)
 {
     std::vector<Vertex> vertices{{
-        {{-1.0f , -1.0f , 0.0f}, {1.0f , 0.0f}} ,
-        {{1.0f , -1.0f , 0.0f}, {0.0f , 0.0f}} ,
-        {{1.0f , 1.0f , 0.0f}, {0.0f , 1.0f}},
-        {{-1.0f , 1.0f , 0.0f}, {1.0f , 1.0f}},
+        {{-0.5f , -0.5f , 0.0f},{1.0f , 0.0f}} ,
+        {{0.5f , -0.5f , 0.0f},{0.0f , 0.0f}} ,
+        {{0.5f , 0.5f , 0.0f},{0.0f , 1.0f}},
+        {{-0.5f , 0.5f , 0.0f},{1.0f , 1.0f}},
     }};
 
     mesh.SetMeshData(vertices, {0, 1, 2, 2, 3, 0});
 }
+
+void Primitives::Sphere(Mesh& mesh, float radius, uint32_t rings, uint32_t sectors)
+{
+    if(radius < 0)
+    {
+        LOG_WARNING("Sphere radius less than 0, sphere will not be generated!");
+        return;
+    }
+
+    if(rings < 4 || sectors < 4)
+    {
+        LOG_WARNING("Rings or sectors are less than 4, sphere will not be generated!");
+        return;
+    }
+
+    std::vector<uint32_t> indices;
+    std::vector<Vertex> vertices;
+
+    uint32_t vectorSize = rings * sectors;
+
+    if(vectorSize > vertices.max_size())
+    {
+        LOG_WARNING("Number of vertices is too big, sphere will not be generated!");
+        return;
+    }
+
+    vertices.resize(vectorSize);
+
+    {
+        float const R = 1.f / static_cast<float>(rings - 1);
+        float const S = 1.f / static_cast<float>(sectors - 1);
+
+        auto vert_iter = vertices.begin();
+        for(uint32_t r = 0; r < rings; ++r)
+        {
+            for(uint32_t s = 0; s < sectors; s++)
+            {
+                float const y = glm::sin(-glm::half_pi<float>() + glm::pi<float>() * r * R);
+                float const x = glm::cos(2 * glm::pi<float>() * s * S) * glm::sin(glm::pi<float>() * r * R);
+                float const z = glm::sin(2 * glm::pi<float>() * s * S) * glm::sin(glm::pi<float>() * r * R);
+                *vert_iter++ = {{x * radius, y * radius, z * radius}, {s * S, r * R}};
+            }
+        }
+    }
+
+    indices.resize(sectors * rings * 6);
+    auto indicesIt = indices.begin();
+    for(uint32_t x = 0; x < sectors; ++x)
+    {
+        float const right = static_cast<float>((x + 1) % sectors);
+        for(uint32_t y = 0; y < rings; ++y)
+        {
+            float const top = static_cast<float>(y);
+            float const bottom = static_cast<float>((y + 1) % rings);
```

Comment:
can all these be `uint32_t`? I don't think you need them to be float for indices